src/app/persistence/tests/TestAttributePersistence.cpp

```diff
@@ -0,0 +1,360 @@
+
+#include <app/AttributeValueDecoder.h>
+#include <app/ConcreteAttributePath.h>
+#include <app/data-model-provider/tests/WriteTesting.h>
+#include <app/persistence/AttributePersistence.h>
+#include <app/persistence/tests/RamAttributePersistenceProvider.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/StringBuilderAdapters.h>
+#include <lib/support/Span.h>
+
+namespace {
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Testing;
+using namespace chip::app::Storage;
+using namespace chip::Testing;
+
+TEST(TestAttributePersistence, TestLoadAndStoreNativeEndian)
+{
+    RamAttributePersistenceProvider ramProvider;
+    AttributePersistence persistence(ramProvider);
+
+    ConcreteAttributePath path(1, 2, 3);
+    ConcreteAttributePath wrongPath(1, 2, 4);
+    constexpr uint32_t kValueToStore = 42;
+    constexpr uint32_t kOtherValue   = 99;
+
+    // Store a fake value
+    {
+        const uint32_t value = kValueToStore;
+        EXPECT_EQ(ramProvider.WriteValue(path, { reinterpret_cast<const uint8_t *>(&value), sizeof(value) }), CHIP_NO_ERROR);
+    }
+
+    // Test loading a value
+    {
+        uint32_t valueRead = 0;
+
+        ASSERT_TRUE(persistence.LoadNativeEndianValue(path, valueRead, kOtherValue));
+        ASSERT_EQ(valueRead, kValueToStore);
+    }
+
+    // Test loading a non-existent value
+    {
+        uint32_t valueRead = 0;
+
+        ASSERT_FALSE(persistence.LoadNativeEndianValue(wrongPath, valueRead, kOtherValue));
+        ASSERT_EQ(valueRead, kOtherValue);
+    }
+
+    // Test loading a removed value
+    {
+        EXPECT_EQ(ramProvider.DeleteValue(path), CHIP_NO_ERROR);
+
+        uint32_t valueRead = 0;
+        ASSERT_FALSE(persistence.LoadNativeEndianValue(path, valueRead, kOtherValue));
+        ASSERT_EQ(valueRead, kOtherValue);
+    }
+}
+
+TEST(TestAttributePersistence, TestLoadAndStoreString)
+{
+    RamAttributePersistenceProvider ramProvider;
+    AttributePersistence persistence(ramProvider);
+    ConcreteAttributePath path(1, 2, 3);
+
+    // Store a fake value
+    {
+        const uint8_t buffer[] = { 5, 'h', 'e', 'l', 'l', 'o' };
+        EXPECT_EQ(ramProvider.WriteValue(path, { buffer, sizeof(buffer) }), CHIP_NO_ERROR);
+    }
+
+    // Test loading a value
+    {
+        char bufferRead[16];
+        ShortPascalString stringRead(bufferRead);
+
+        ASSERT_TRUE(persistence.Load(path, stringRead, std::nullopt));
+        ASSERT_TRUE(stringRead.Content().data_equal("hello"_span));
+    }
+
+    // Test loading a non-existent value
+    {
+        ConcreteAttributePath wrongPath(1, 2, 4);
+        char bufferRead[16];
+        ShortPascalString stringRead(bufferRead);
+
+        ASSERT_FALSE(persistence.Load(wrongPath, stringRead, "default"_span));
+        ASSERT_TRUE(stringRead.Content().data_equal("default"_span));
+    }
+}
+
+TEST(TestAttributePersistence, TestNativeRawValueViaDecoder)
+{
+    RamAttributePersistenceProvider ramProvider;
```

Comment:
Curious why we are not using the `TestPersistentStorageDelegate.h` with an early impl fo the persistence storage instead of having a separate impl in RAM of the API?

Reply:
Updated. This should give us more coverage now that I touch the default implementation.